doc/checknotes.1
.TH CHECKNOTES 1 "August 2004" "Newts" "Newts Reference Manual"

.SH NAME
checknotes \- check for new and updated notes

.SH SYNOPSIS
.B checknotes
[\fIoptions\fR] [\fINOTESFILE\fR]...

.SH DESCRIPTION
.B checknotes
checks various notesfiles to see if they've been updated since the last time
you read through them.

More specifically,
.B checknotes
compares a notesfile's modification time to the time stored in the sequencer
for that notesfile.  If the notesfile's modification time is more recent, the
notesfile is marked as having new material to read.

.SH OPTIONS

.TP
\fB\-a\fR, \fB\-\^\-alternate\fR=\fINAME\fR
Use a named subsequencer.  Each subsequencer has its own independent set of
timestamps, and each can be used independently to sequence through the same
notesfiles.

This option was originally used for shared accounts, so multiple users could
track their sequencers individually.  Since most sites have moved away from
using shared accounts, this option is rarely used.

.TP
\fB\-f\fR, \fB\-\^\-file\fR=\fIFILE\fR
Read a list of notesfiles to check from the specified file, instead of
expecting such a list to be provided on the command line.

.TP
\fB\-h\fR, \fB\-\^\-help\fR
Print a summary of usage and command-line options for
.B checknotes
and exit.

.TP
\fB\-q\fR, \fB\-s\fR, \fB\-\^\-quiet\fR, \fB\-\^\-silent\fR
Do not output any messages, regardless of whether there are notesfiles with new
or updated material.  This is primarily useful in a shell script, where the
error code (0 if there are new notes, nonzero if there are no new notes or an
error occured) can be used effectively.

.TP
\fB\-v\fR, \fB\-\^\-verbose\fR
Print a message for each notesfile with new or updated notes.

.TP
\fB\-\^\-debug\fR
Print debugging messages to standard error.

.TP
\fB\-\^\-version\fR
Print version information for
.B checknotes
and exit.

.SH AUTHOR
.B Newts
was written and is maintained by Tyler Berry <tyler+newts@thoughtlocker.net>.

.SH SEE ALSO
\fBautoseq\fR(1), \fBgetnote\fR(1), \fBmknf\fR(1), \fBnfadmin\fR(1),
\fBnfdump\fR(1), \fBnfload\fR(1), \fBnfmail\fR(1), \fBnfpipe\fR(1),
\fBnfprint\fR(1), \fBnfstats\fR(1), \fBnftimestamp\fR(1), \fBnotes\fR(1),
\fBrmnf\fR(1)

The full documentation for
.B Newts
is maintained as a Texinfo manual.  If the
.B info
and
.B Newts
programs are properly installed at your site, the command
.IP
.B info newts
.PP
should give you access to the complete manual.


NEWS
Newts NEWS                                                         -*- Text -*-
-------------------------------------------------------------------------------

* ______ - Newts 0.15

- More unit tests.
- More changes and additions to the client API.
- More documentation of the client API.

* Wednesday, December 14, 2005 - Newts 0.14.8

- Include the SVN revision number in the output of --version.
- Added unit tests for the client API.
- Began adding documentation for the client API in a Doxygen-compatible way.
- Substantial changes and additions to the client API.
- Substantial refactoring in the main client, both to comply with the revised
  client API and to clean the code up in general.
- Modified the utility frontend programs to use the revised client API.
- Refactored the pager to be modular, for purposes of future rewrites to add
  things like Unicode support.
- Use gnulib-tool to generate Gnulib support.
- Automatically use the most recent version of Libtool.

* Tuesday, July 26, 2005 - Newts 0.14.7

- Fixed signal handling so that suspending a spawned process works correctly.
- Added David Mellis's python bindings to the repository and distribution.

* Friday, June 16, 2005 - Newts 0.14.6

- Fixed a bug in nfdump where the last note in a dumped notesfile was omitted.
- Added 'g' to perform one-touch seqing.

* Monday, March 21, 2005 - Newts 0.14.5

- Updated to Libtool 1.5.14.
- Fixed a segfault when trying to use the alternate sequencer.
- Made the alternate sequencer activate the sequencer automatically.
- Made the alternate sequencer work correctly with the '--user' option.
- Fixed the broken --time option, which was not actually setting the time at
  all. Oops.
- Fixed a few odd display bugs.
- Display messages are now 'Unicode-friendly'.
  See <http://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html> for details.
- Resynced with Gnulib.

* Sunday, January 23, 2005 - Newts 0.14.4

- Updated to Libtool 1.5.10.
- Corrected some mistakes in the Texinfo manual.
- Fixed a bug which would cause a segfault when replying to a note from a local
  user who had been deleted from the system, or forwarding a mail from a note
  written by such a user.
- Fixed a bug in nfload where the dumpfile would fail to open if nfload was run
  as root. (It was dropping root permissions and not picking them back up.)
- Resynced with Gnulib, which involved some fairly major updates to xmalloc and
  friends, as the interface Gnulib uses changed.

* Tuesday, September 14, 2004 - Newts 0.14.3

- Corrected a typo in nfadmin.1.
- Synchronized with new files in Gnulib.

* Tuesday, August 10, 2004 - Newts 0.14.2

- Corrected a Whole Bunch of inconsistencies between the Texinfo manual, the
  newly-developed manpages, and the actual operation of the programs.
- Added manpages for all currently-built programs.
- Renamed 'printnote' to 'getnote'.
- Removed the '-d' short option from all programs, which used to correspond to
  '--debug'.  It was blocking the use of '-d' for more important options which
  are far more likely to see use.

* Tuesday, June 29, 2004 - Newts 0.14.1

- Completely redid all compatibility code using a recent CVS version of
  Gnulib.

* Thursday, June 10, 2004 - Newts 0.14

- Adding the missing code to the nfload program which will load access entries
  from the dumpfile into the notesfile.
- Added debugging support to all of the various binaries built by the package,
  via the --debug command-line option.
- Added an option for debugging that allows the notes client to setuid to an
  arbitrary user when being run by root. Useful for running notes inside gdb.
- Changed the license for the Newts Reference Manual - it's no longer GFDL.
- Fixed a bug which would cause a segfault when reading a note from a local
  user who had been deleted from the system.
- Reorganized the entire directory structure of the source tree.
- Reorganized header files to install the relevant interfaces.
- Began systematic documentation of the libnewtsclient interface.

* Tuesday, March 9, 2004 - Newts 0.13.2

- Fixed a display bug which caused informational messages and prompts to fail
  to display when a person tried to enter a notesfile without rights.
- Syncronized with new files in Gnulib.

* Friday, January 16, 2003 - Newts 0.13.1

- Synchronized with new releases of GNU Autotools and Gnulib.

* Tuesday, December 9, 2003 - Newts 0.13

- Added nfdump, completing the set of basic utilities.
- Added some sanity checks to the 't' talk command.
- Various small bits of code cleanup.
- Made sure that root doesn't accidentally lose the ability to write files.

* Monday, September 29, 2003 - Newts 0.12.7

- Portability work for sparc-sun-solaris2.6.
- utmp scanning to determine whether an author is online.
- Synchronization and updates for portability code.

* Thursday, September 18, 2003 - Newts 0.12.6

- Added a couple missing m4 files.

* Friday, September 5, 2003 - Newts 0.12.5

- Code cleanup, fixes, and error-checking.

* Saturday, August 30, 2003 - Newts 0.12.4

- Added man pages for notes(1) and autoseq(1).
- Fixed inconsistencies between documentation and actual behavior.
- Made a couple command line options make more sense.
- Fixed distribution of a couple missing files.

* Tuesday, August 26, 2003 - Newts 0.12.3

- Fixed a mistake with an earlier fix to the 'N' key.
- Prepared for a CVS branch to add Readline support.

* Tuesday, August 19, 2003 - Newts 0.12.2

- Interim release for code maintenance.
- Lots and lots of code cleanup.
- Fixed lots of FIXMEs in the code; these were largely for incorrect usage of
  one kind or another; static buffers instead of dynamic alloc, not checking
  return values, etc. It's not there yet, but it's getting there.

* Wednesday, August 13, 2003 - Newts 0.12.1

- Staving off disaster, a bug in notesfile parsing was discovered and slain in
  the embryonic version 0.12, which is supplanted by this release.
- Finally fixed that nfload bug that's been driving me nuts for the past three

clients/frontends/checknotes.c
/*
 * checknotes.c - check sequencer to see if there are any new notes
 *
 * This file is part of the Newts notesfile system.
 * Copyright (C) 2003, 2004, 2005 Tyler Berry.
 *
 * Based in part on checknotes.c from the UIUC notes distribution by Ray Essick
 * and Rob Kolstad.  Any work derived from this source code is required to
 * retain this notice.
 *
 * Newts is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Newts is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Newts; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "frontend.h"

#include "dirname.h"
#include "getopt.h"
#include "strtok_r.h"

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

enum
  {
    NORMAL,
    SILENT,
    VERBOSE
  };

/* Whether to display debugging messages. */
int debug = FALSE;

/* Whether or not to ignore the blacklist. */
int no_blacklist = FALSE;

/* The sequencer time we're dealing with here. */
time_t seqtime;

/* Placeholder for the blacklist code. */
const short white_basenotes = FALSE;

static int verify_sequencer (struct notesfile *nf);

int
main (int argc, char **argv)
{
  List nflist;
  struct notesfile *nf;

  int fileflag = FALSE;
  int updated = 0;
  short verbosity = NORMAL;
  char *seqname;

  int opt;
  int option_index = 0;

  struct option long_options[] =
    {
      {N_("alternate"),1,0,'a'},
      {N_("debug"),0,0,'D'},
      {N_("file"),1,0,'f'},
      {N_("quiet"),0,0,'s'},
      {N_("silent"),0,0,'s'},
      {N_("verbose"),0,0,'v'},
      {N_("no-blacklist"),0,0,'z'},
      {N_("help"),0,0,'h'},
      {N_("version"),0,0,0},
      {0,0,0,0}
    };

#ifdef __GLIBC__
  program_name = program_invocation_short_name;
#else
  program_name = base_name (argv[0]);
#endif

  /* Initialize i18n. */

#ifdef HAVE_SETLOCALE
  setlocale (LC_ALL, "");
#endif

#if ENABLE_NLS
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  setup ();
  seqname = newts_strdup (username);
  list_init (&nflist,
             (void * (*) (void)) nfref_alloc,
             (void (*) (void *)) nfref_free,
             NULL);

  while ((opt = getopt_long (argc, argv, N_("a:f:hnqsv"),
                             long_options, &option_index)) != -1)
    {
      switch (opt)
        {
        case 0:
          {
            printf_version_string (N_("checknotes"));

            newts_free (seqname);
            list_destroy (&nflist);
            teardown ();

            if (fclose (stdout) == EOF)
              error (1, errno, _("error writing output"));
            exit (0);
          }

        case 'a':
          newts_nrealloc (seqname, strlen (username) + strlen (optarg) + 2,
                          sizeof (char));
          strcpy (seqname, username);
          strcat (seqname, ":");
          strcat (seqname, optarg);
          break;

        case 'D':
          debug = TRUE;
          break;

        case 'f':
          if (parse_file (optarg, &nflist))
            fileflag = TRUE;
          break;

        case 'n':
          verbosity = NORMAL;
          break;

        case 'q':
          verbosity = NORMAL;
          break;

        case 's':
          verbosity = SILENT;
          break;

        case 'v':
          verbosity = VERBOSE;
          break;

        case 'z':
          no_blacklist = TRUE;
          break;

        case 'h':
          printf (_("Usage: %s [OPTION]... [NOTESFILE]...\n"
                    "Check for new notes in your sequencer path or in listed notesfiles.\n\n"),
                    program_name);

          printf (_("  -a, --alternate=SEQ   Use alternate sequencer SEQ\n"
                    "  -f, --file=FILE       Read list of notesfiles to view from specified file\n"
                    "  -s, --silent          Display no output\n"
                    "  -v, --verbose         Display each notesfile with new notes\n"
                    "  -z, --no-blacklist    Ignore blacklist while checking notesfiles.\n"
                    "      --debug           Display debugging messages\n"
                    "      --quiet           Same as -s\n\n"
                    "  -h, --help            Display this help and exit\n"
                    "      --version         Display version information and exit\n\n"));

          printf (_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);

          newts_free (seqname);
          list_destroy (&nflist);
          teardown ();

          if (fclose (stdout) == EOF)
            error (1, errno, _("error writing output"));

          exit (1); /* Because exit value 0 is significant. */

        case '?':
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);

          newts_free (seqname);
          list_destroy (&nflist);
          teardown ();

          if (fclose (stdout) == EOF)
            error (1, errno, _("error writing output"));

          exit (1);
        }
    }

  if (optind == argc && !fileflag)
    {
      char *copy, *list, *token, *nfseq;

      nfseq = getenv ("NFSEQ");
      if (nfseq == NULL)
        {
          fprintf (stderr, _("%s: NFSEQ environment variable not set\n"),
                   program_name);
          fprintf (stderr, _("See 'info newts' for more information.\n"));
          exit (1);
        }
      copy = newts_strdup (nfseq);

      token = strtok_r (copy, ", ", &list);
      while (token != NULL)
        {
          parse_nf (token, &nflist);
          token = strtok_r (NULL, ", ", &list);
        }

      newts_free (copy);
    }
  else
    {
      while (optind < argc)
        parse_nf (argv[optind++], &nflist);
    }

  {
    ListNode *node = list_head (&nflist);

    while (node != NULL)
      {
        int error = open_nf ((newts_nfref *) list_data (node), nf);

        node = list_next (node);

        if (error != NEWTS_NO_ERROR)
          {
            fprintf (stderr, _("%s: error opening '%s'\n"), program_name,
                     nfref_pretty_name (nf->ref));
            continue;
          }

        get_seqtime (nf->ref, seqname, &seqtime);

        if (difftime (seqtime, nf->modified) <= 0 &&
            verify_sequencer (nf))
          {
            updated++;
            if (verbosity == VERBOSE)
              {
                printf (N_("%s\n"), nfref_pretty_name (nf->ref));
              }
          }
      }
  }

  if (verbosity == NORMAL)
    {
      if (updated)
        printf (_("There are new notes\n"));
      else
        printf (_("There are no new notes\n"));
    }

  newts_free (seqname);
  list_destroy (&nflist);
  teardown ();

  if (fclose (stdout) == EOF)
    error (1, errno, _("error writing output"));

  exit (updated ? 0 : 1);
}

/* verify_sequencer - look through a notesfile with the sequencer to see if it
 * really has new notes, with our blacklist taken into effect.
 *
 * Returns: TRUE if the notesfile really has new notes, FALSE otherwise.
 */

static int
verify_sequencer (struct notesfile *nf)
{
  struct newt note;
  memset (&note, 0, sizeof (struct newt));
  nfref_copy (&note.nr.nfr, nf->ref);
  note.nr.notenum = 0;
  note.nr.respnum = 0;

  note.nr.notenum = get_next_note (&note.nr, seqtime);

  if (note.nr.notenum < 0)
    return FALSE;

  while (TRUE)
    {
      get_note (&note, FALSE);

      if (difftime (note.modified, seqtime) > 0 &&
          (difftime (note.created, seqtime) > 0 || note.nr.respnum != 0))
        {
          if (!(euid == note.auth.uid && strcmp (fqdn, note.auth.system) == 0)
              && !blacklisted (&note))
            return TRUE;
        }

      if (note.nr.respnum != note.total_resps)
        {
          if ((note.nr.respnum = get_next_resp (&note.nr, seqtime)) > 0)
            continue;
        }

      note.nr.respnum = 0;
      if ((note.nr.notenum = get_next_note (&note.nr, seqtime)) > 0)
        continue;
      else
        return FALSE;
    }
}
